DOCSUP-6543: Edit and translate Show Clusters and MergeTree (#24130)
ClickHouse/ClickHouse: docs/ru/operations/settings/merge-tree-settings.md

# Настройки MergeTree таблиц {#merge-tree-settings}

> Значения настроек всех MergeTree таблиц собраны в таблице `system.merge_tree_settings`. Их можно переопределить в разделе `merge_tree` файла `config.xml`, а задать — в секции `SETTINGS` каждой таблицы.

Reviewer: ```suggestion
Значения настроек всех MergeTree таблиц собраны в таблице `system.merge_tree_settings`. Их можно переопределить в разделе `merge_tree` файла `config.xml` или задать в секции `SETTINGS` каждой таблицы.
```

## parts_to_throw_insert {#parts-to-throw-insert}

Eсли число активных кусков в партиции больше значения `parts_to_throw_insert`, то INSERT прерывается с исключением: `Too many parts (N). Merges are processing significantly slower than inserts`.

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 300.

Чтобы производительность запросов `SELECT` стала максимальной, необходимо минимизировать количество обрабатываемых кусков, см. [Дизайн MergeTree](../../development/architecture.md#merge-tree).

> Можно установить значение побольше — 600 (1200) кусков. Тогда ошибка `Too many parts` будет появляться реже, но при этом могут возникнуть проблемы с фоновыми слияниями и производительностью `SELECT`-запросов.

Reviewer: ```suggestion
Можно установить значение больше — 600 (1200) кусков. Тогда ошибка `Too many parts` будет появляться реже, но при этом могут возникнуть проблемы с фоновыми слияниями и производительностью `SELECT`-запросов.
```

Величина задержки (в миллисекундах) для `INSERT` вычисляется по формуле:

```code
max_k = parts_to_throw_insert - parts_to_delay_insert
k = 1 + parts_count_in_partition - parts_to_delay_insert
delay_milliseconds = pow(max_delay_to_insert * 1000, k / max_k)
```

Т.е. если в партиции уже 299 кусков и parts_to_throw_insert = 300, parts_to_delay_insert = 150, а max_delay_to_insert = 1, то `INSERT` замедлится на `pow( 1 * 1000, (1 + 299 - 150) / (300 - 150) ) = 1000` миллисекунд.

## max_parts_in_total {#max-parts-in-total}

Eсли суммарное число активных кусков во всех партициях таблицы больше значения `max_parts_in_total`, то INSERT прерывается с исключением `Too many parts (N)`.

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 100000.

> С большим числом кусков в таблице производительность запросов ClickHouse снижается, а время старта ClickHouse — увеличивается. Чаще всего, это — следствие неправильного дизайна (ошибки выбора стратегии партиционирования, например, слишком мелкие партиции).

Reviewer: ```suggestion
С большим числом кусков в таблице производительность запросов ClickHouse снижается, а время старта ClickHouse — увеличивается. Чаще всего это следствие неправильного дизайна (ошибки выбора стратегии партиционирования, например, слишком мелкие партиции).
```

## replicated_deduplication_window {#replicated-deduplication-window}

> Количество хеш-сумм вставленных последними блоков, которые хранятся в Zookeeper.

Reviewer: ```suggestion
Количество хеш-сумм последних вставленных блоков, которые хранятся в Zookeeper.
```

Возможные значения:

-   Положительное целое число.

> 

Reviewer: -   0 (без ограничений)

Значение по умолчанию: 100.

> Команда `Insert` создает один или несколько блоков (кусков). При вставке в Replicated таблицы ClickHouse для [дедупликации вставок](../../engines/table-engines/mergetree-family/replication.md) записывает в Zookeeper хеш-суммы созданных кусков. Но хранятся только последние хеш-суммы кусков `replicated_deduplication_window`. Самые старые хеш-суммы удаляются из Zookeeper.

Reviewer: ```suggestion
Команда `Insert` создает один или несколько блоков (кусков). При вставке в Replicated таблицы ClickHouse для [дедупликации вставок](../../engines/table-engines/mergetree-family/replication.md) записывает в Zookeeper хеш-суммы созданных кусков. Но хранятся только последние `replicated_deduplication_window` хеш-сумм. Самые старые хеш-суммы удаляются из Zookeeper.
```

> Большое число `replicated_deduplication_window` тормозит `Insert`. Хеш-сумма рассчитывается по объединению имен и типов полей, а также данных вставленного куска (потока байт).

Reviewer: ```suggestion
Большое значение `replicated_deduplication_window` замедляет `Insert`, так как приходится сравнивать большее количество хеш-сумм. 
Хеш-сумма рассчитывается по объединению имен и типов полей, а также данных вставленного куска (потока байт).
```

## replicated_deduplication_window_seconds {#replicated-deduplication-window-seconds}

> Время хранения (в секундах) хеш-сумм вставленных блоков в Zookeeper.

Reviewer: ```suggestion
Время хранения (в секундах) хеш-сумм вставленных блоков в Zookeeper. По истечении этого времени хеш-суммы удаляются.
```

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 604800 (1 неделя).

> Аналогично значение [replicated_deduplication_window](#replicated-deduplication-window) задает время хранения хеш-сумм блоков для дедупликции `Insert`. Хеш-суммы старше значения `replicated_deduplication_window_seconds` удаляются из Zookeeper, даже если их меньше чем `replicated_deduplication_window`.

Reviewer: ```suggestion
Аналогично [replicated_deduplication_window](#replicated-deduplication-window), настройка `replicated_deduplication_window_seconds` задает время хранения хеш-сумм блоков для дедупликции `Insert`. Хеш-суммы старше значения `replicated_deduplication_window_seconds` удаляются из Zookeeper, даже если количество оставшихся хеш-сумм станет меньше чем `replicated_deduplication_window`.
```

## old_parts_lifetime {#old-parts-lifetime}

> Время (в секундах) хранения неактивных кусков для защиты от потери данных при спонтанной перезагрузке сервера или ОС.

Reviewer: ```suggestion
Время (в секундах) хранения неактивных кусков для защиты от потери данных при спонтанной перезагрузке сервера.
```

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 480.

> После объединения нескольких кусков в один новый происходит следующее: ClickHouse помечает исходные куски как неактивные и удаляет их по прошествию `old_parts_lifetime` секунд.

Reviewer: ```suggestion
После объединения нескольких кусков в один новый ClickHouse помечает исходные куски как неактивные и удаляет их по прошествии `old_parts_lifetime` секунд.
```

Неактивные куски удаляются, если они не нужны для текущих запросов, т.е. если счетчик ссылок куска `refcount` имеет нулевое значение.

Неактивные куски удаляются со временем, потому что `fsync` не вызывается при записи нового куска. Получается, что новый кусок некоторое время находится только в оперативной памяти сервера. Если сервер вдруг перезагружается, новый (слитый) кусок может исчезнуть или испортиться. Тогда ClickHouse обнаружит проблему на старте во время проверки целостности кусков, вернет неактивные куски в список активных и снова сольет их позже. В этом случае испорченный кусок получает новое имя (добавляется префикс broken_) и попадает в папку `detached`. Если проверка целостности не выявляет проблем в слитом куске, то исходные неактивные куски переименовываются (добавляется префикс ignored_) и перемещаются в папку `detached`.

> Стандартное значение `Linux dirty_expire_centisecs` — 30 секунд. Это максимальное время, в течение которого записанные данные хранятся только в оперативной памяти. Если нагрузка на дисковую систему большая, то данные записываются намного позже. Эксперименты показали 480 секунд — это время, за которое новый кусок гарантированно запишется на диск.

Reviewer: ```suggestion
Стандартное значение `dirty_expire_centisecs` — 30 секунд. Это максимальное время, в течение которого записанные данные хранятся только в оперативной памяти. Если нагрузка на дисковую систему большая, то данные записываются намного позже. Эксперименты показали 480 секунд — это время, за которое новый кусок гарантированно запишется на диск.
```

Author: @olgarev Оля, а может тут стоит заменить "было подобрано" на "подобрали"?

## max_bytes_to_merge_at_max_space_in_pool {#max-bytes-to-merge-at-max-space-in-pool}

Максимальный суммарный размер кусков (в байтах) в одном слиянии, если есть свободные ресурсы в фоновом пуле.

> `max_bytes_to_merge_at_max_space_in_pool` — примерно соответствует максимально возможному размеру куска, созданного автоматическим фоновым слиянием.

Reviewer: ```suggestion
`max_bytes_to_merge_at_max_space_in_pool` примерно соответствует максимально возможному размеру куска, созданного автоматическим фоновым слиянием.
```

Возможные значения:

-   Положительное целое число.

> Значение по умолчанию: 161061273600 (150ГБ).

Reviewer: ```suggestion
Значение по умолчанию: 161061273600 (150Гб).
```

> Планировщик слияний периодически анализирует размер и количество кусков в партициях. Если хватает ресурсов, то начинает фоновое слияние в фоновом пуле. Слияния происходят до тех пор, пока суммарный размер входных кусков не достигнет `max_bytes_to_merge_at_max_space_in_pool`.

Reviewer: ```suggestion
Планировщик слияний периодически анализирует размер и количество кусков в партициях, и если в пуле хватает ресурсов, то начинает фоновое слияние. Слияния выполняются до тех пор, пока суммарный размер входных кусков не достигнет `max_bytes_to_merge_at_max_space_in_pool`.
```

> Слияния, начатые по `optimize final`, не учитывают `max_bytes_to_merge_at_max_space_in_pool` и сливают куски, используя только доступные ресурсы (например, свободное дисковое пространство) до тех пор, пока в партиции не останется один кусок.

Reviewer: ```suggestion
Слияния, начатые по [OPTIMIZE FINAL](../../sql-reference/statements/optimize.md), не учитывают `max_bytes_to_merge_at_max_space_in_pool` и объединяют куски пока есть доступные ресурсы (свободное дисковое пространство) до тех пор, пока в партиции не останется один кусок.
```

## max_bytes_to_merge_at_min_space_in_pool {#max-bytes-to-merge-at-min-space-in-pool}

Максимальный суммарный размер кусков (в байтах) в одном слиянии при минимуме свободных ресурсов в фоновом пуле.

Возможные значения:

-   Положительное целое число.

> Значение по умолчанию: 1048576

Reviewer: ```suggestion
Значение по умолчанию: 1048576 (1 МБ).
```

> `max_bytes_to_merge_at_min_space_in_pool` задает максимальный суммарный размер кусков, которые можно объединить несмотря на нехватку свободных ресурсов в фоновом пуле (дискового пространства). Это нужно, чтобы уменьшить количество маленьких кусков и снизить вероятность ошибки `Too many parts`.

Reviewer: ```suggestion
`max_bytes_to_merge_at_min_space_in_pool` задает максимальный суммарный размер кусков, которые можно объединить несмотря на нехватку свободных ресурсов (дискового пространства) в фоновом пуле. Это нужно, чтобы уменьшить количество маленьких кусков и снизить вероятность ошибки `Too many parts`.
```

> Слияния резервируют дисковое пространство, удваивая суммарный размер кусков в слиянии. Если свободного места на диске мало, то оно может быть уже зарезервировано текущими слиянияними. Поэтому другие слияния не начинаются, а количество маленьких кусков в партиции растет с каждым `Insert`.

Reviewer: ```suggestion
Слияния резервируют дисковое пространство, удваивая суммарный размер кусков в слиянии. Поэтому при малом объеме свободного места на диске может сложиться ситуация, когда свободное место есть, но оно уже зарезервировано текущими слияниями. Из-за этого другие слияния не начинаются, а количество маленьких кусков в партиции растет с каждым запросом `INSERT`.
```

## merge_max_block_size {#merge-max-block-size}

Количество строк в блоках, которые читаются из объединяемых кусков.

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 8192

> Слияние читает строки из кусков блоками по `merge_max_block_size` строк, производит слияние и пишет результат в новый кусок. Читаемый блок помещается в оперативную память, т.е. `merge_max_block_size` влияет на размер оперативной памяти, необходимой для слияния. Таким образом, слияния могут потреблять большое количество оперативной памяти для таблиц, хранящих очень большие строки (если средний размер строки 100кб, а при слиянии 10 кусков будет использовано (100кб * 10 * 8192) =~ 8ГБ ОЗУ). Чтобы сократить необходимую оперативную память, нужно уменьшить `merge_max_block_size`.

Reviewer: ```suggestion
Слияние читает строки из кусков блоками по `merge_max_block_size` строк, производит слияние и записывает результат в новый кусок. Читаемый блок помещается в оперативную память, т.е. `merge_max_block_size` влияет на размер оперативной памяти, необходимой для слияния. Таким образом, слияния могут потреблять большое количество оперативной памяти для таблиц, хранящих очень большие строки (если средний размер строки 100кб, то при слиянии 10 кусков будет использовано (100кб * 10 * 8192) =~ 8ГБ оперативной памяти). Уменьшив `merge_max_block_size`, можно сократить размер оперативной памяти, необходимой для слияния.
```

Время хранения (в секундах) хеш-сумм вставленных блоков в Zookeeper. По истечении этого времени хеш-суммы удаляются.

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 604800 (1 неделя).

Аналогично [replicated_deduplication_window](#replicated-deduplication-window), настройка `replicated_deduplication_window_seconds` задает время хранения хеш-сумм блоков для дедупликции `Insert`. Хеш-суммы старше значения `replicated_deduplication_window_seconds` удаляются из Zookeeper, даже если количество оставшихся хеш-сумм станет меньше чем `replicated_deduplication_window`.

## old_parts_lifetime {#old-parts-lifetime}

Время (в секундах) хранения неактивных кусков для защиты от потери данных при спонтанной перезагрузке сервера.

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 480.

После объединения нескольких кусков в один новый ClickHouse помечает исходные куски как неактивные и удаляет их по прошествии `old_parts_lifetime` секунд.
Неактивные куски удаляются, если они не нужны для текущих запросов, т.е. если счетчик ссылок куска `refcount` имеет нулевое значение.

> Неактивные куски удаляются со временем, потому что `fsync` не вызывается при записи нового куска. Получается, что новый кусок некоторое время находится только в оперативной памяти сервера. Если сервер вдруг перезагружается, новый (слитый) кусок может исчезнуть или испортиться. Тогда ClickHouse обнаружит проблему на старте во время проверки целостности кусков, вернет неактивные куски в список активных и снова сольет их позже. В этом случае испорченный кусок получает новое имя (добавляется префикс broken_) и попадает в папку `detached`. Если проверка целостности не выявляет проблем в слитом куске, то исходные неактивные куски переименовываются (добавляется префикс ignored_) и перемещаются в папку `detached`.

Reviewer: ```suggestion
Во время запуска сервер ClickHouse проверяет целостность кусков. 
Если новый (слитый) кусок поврежден, ClickHouse возвращает неактивные куски в список активных и позже снова выполняет слияние. В этом случае испорченный кусок получает новое имя (добавляется префикс `broken_`) и попадает в каталог `detached`. 
Если проверка целостности не выявляет проблем в слитом куске, то исходные неактивные куски переименовываются (добавляется префикс `ignored_`) и перемещаются в каталог `detached`.
```

## old_parts_lifetime {#old-parts-lifetime}

Время (в секундах) хранения неактивных кусков, для защиты от потери данных при спонтанной перезагрузке сервера или О.С.
Время (в секундах) хранения неактивных кусков для защиты от потери данных при спонтанной перезагрузке сервера.

Возможные значения:

-   Положительное целое число.

Значение по умолчанию: 480.

После слияния нескольких кусков в новый кусок, ClickHouse помечает исходные куски как неактивные и удаляет их после `old_parts_lifetime` секунд.
Неактивные куски удаляются, если они не используются в текущих запросах, т.е. если счетчик ссылок куска – `refcount` равен нулю.
После объединения нескольких кусков в один новый ClickHouse помечает исходные куски как неактивные и удаляет их по прошествии `old_parts_lifetime` секунд.
Неактивные куски удаляются, если они не нужны для текущих запросов, т.е. если счетчик ссылок куска `refcount` имеет нулевое значение.

> Неактивные куски удаляются не сразу, потому что при записи нового куска не вызывается `fsync`, т.е. некоторое время новый кусок находится только в оперативной памяти сервера (кеше О.С.). Т.о. при спонтанной перезагрузке сервера новый (смерженный) кусок может быть потерян или испорчен. В этом случае ClickHouse в процессе старта при проверке целостности кусков обнаружит проблему, вернет неактивные куски в список активных и позже заново их смержит. Сломанный кусок в этом случае переименовывается (добавляется префикс broken_) и перемещается в папку detached. Если проверка целостности не обнаруживает проблем в смерженном куске, то исходные неактивные куски переименовываются (добавляется префикс ignored_) и перемещаются в папку detached.

Reviewer 2: Мне кажется, что предложение про fsync следует оставить - хотя бы в качестве примера возможной причины, почему новый кусок может оказаться битым.

После объединения нескольких кусков в один новый ClickHouse помечает исходные куски как неактивные и удаляет их по прошествии `old_parts_lifetime` секунд.
Неактивные куски удаляются, если они не нужны для текущих запросов, т.е. если счетчик ссылок куска `refcount` имеет нулевое значение.

> Неактивные куски удаляются не сразу, потому что при записи нового куска не вызывается `fsync`, т.е. некоторое время новый кусок находится только в оперативной памяти сервера (кеше О.С.). Т.о. при спонтанной перезагрузке сервера новый (смерженный) кусок может быть потерян или испорчен. 

Reviewer 2: Я думаю, что здесь стóит привести к такому же стилю, как и остальные новые части текста вокруг :)

-   Положительное целое число.

> Значение по умолчанию: `auto` (количество ядер процессора).

Reviewer: ```suggestion
Значение по умолчанию: определяется автоматически (по количеству ядер процессора).
```

> На старте ClickHouse читает все куски из всех таблиц (файлы с метаданными кусков), чтобы построить в ОЗУ список всех кусков. В некоторых системах с большим количеством кусков этот процесс бывает долгим. Это время уменьшается за счет увеличения `max_part_loading_threads` (если при этом хватает ресурсов CPU и диска).

Reviewer: ```suggestion
На старте ClickHouse читает все куски из всех таблиц (читает файлы с метаданными кусков), чтобы построить в оперативной памяти список всех кусков. В некоторых системах с большим количеством кусков этот процесс может занимать длительное время. Это время можно сократить, увеличив `max_part_loading_threads` (если при этом хватает ресурсов процессора и диска).
```
